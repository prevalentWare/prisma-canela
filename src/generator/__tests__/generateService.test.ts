import { generateServiceFileContent } from '@generator/generateService';
import { describe, it, expect, vi } from 'vitest';
import type { ParsedModel } from '@parser/types';
import type { DMMF } from '@prisma/generator-helper';

// Mock data for a simple User model
const userModel: ParsedModel = {
  name: 'User',
  fields: [
    {
      name: 'id',
      type: 'string',
      kind: 'scalar',
      isId: true,
      isRequired: true,
      isUnique: true,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: 'email',
      type: 'string',
      kind: 'scalar',
      isId: false,
      isRequired: true,
      isUnique: true,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: 'name',
      type: 'string',
      kind: 'scalar',
      isId: false,
      isRequired: false,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: 'role',
      type: 'enum',
      kind: 'enum',
      enumName: 'Role',
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    }, // Example enum field
  ],
  dbName: null,
};

// Mock data for a model without an ID field (e.g., LogEntry)
const logEntryModel: ParsedModel = {
  name: 'LogEntry',
  fields: [
    {
      name: 'message',
      type: 'string',
      kind: 'scalar',
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: 'level',
      type: 'string',
      kind: 'scalar',
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: 'timestamp',
      type: 'date',
      kind: 'scalar',
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
  ],
  dbName: null,
};

// Mock getPrismaPath to return a fixed path
vi.mock('@generator/getPrismaPath', () => ({
  getPrismaPath: vi.fn().mockResolvedValue('prisma/schema/generated/client'),
}));

describe('generateServiceFileContent', () => {
  it('should generate correct service file content for a User model', async () => {
    const model: ParsedModel = {
      name: 'User',
      dbName: null,
      fields: [
        {
          name: 'id',
          type: 'string',
          kind: 'scalar' as DMMF.FieldKind,
          isRequired: true,
          isId: true,
          isUnique: true,
          isUpdatedAt: false,
          isList: false,
          hasDefaultValue: true,
        },
        {
          name: 'email',
          type: 'string',
          kind: 'scalar' as DMMF.FieldKind,
          isRequired: true,
          isId: false,
          isUnique: true,
          isUpdatedAt: false,
          isList: false,
          hasDefaultValue: false,
        },
      ],
    };

    const result = await generateServiceFileContent(model);
    expect(result).toMatchSnapshot();
    expect(result).toContain('export const findManyUser');
    expect(result).toContain('export const createUser');
    expect(result).toContain('export const findUserById');
    expect(result).toContain('export const updateUser');
    expect(result).toContain('export const deleteUser');

    // Model name in prisma client access should be camelCase
    expect(result).toContain('prisma.user.findMany');
    expect(result).toContain('prisma.user.create');
    expect(result).toContain('prisma.user.findUnique');
    expect(result).toContain('prisma.user.update');
    expect(result).toContain('prisma.user.delete');
  });

  it('should generate service file content with a custom prisma client path', async () => {
    const result = await generateServiceFileContent(
      userModel,
      '../../libs/prisma'
    );
    expect(result).toContain("from '../../libs/prisma'");
    // Use snapshot testing for the generated code string
    expect(result).toMatchSnapshot();
  });

  it('should generate service file content for a model without an ID', async () => {
    const model: ParsedModel = {
      name: 'Junction',
      dbName: null,
      fields: [
        {
          name: 'userId',
          type: 'relation',
          kind: 'object' as DMMF.FieldKind,
          isRequired: true,
          isId: false,
          isUnique: false,
          isUpdatedAt: false,
          isList: false,
          hasDefaultValue: false,
          relationInfo: {
            relatedModelName: 'User',
            relationName: null,
          },
        },
      ],
    };

    const result = await generateServiceFileContent(model);
    expect(result).toMatchSnapshot();
    // Should only include findMany and create functions
    expect(result).toContain('export const findManyJunction');
    expect(result).toContain('export const createJunction');
    // Should not include ID-based functions
    expect(result).not.toContain('export const findJunctionById');
    expect(result).not.toContain('export const updateJunction');
    expect(result).not.toContain('export const deleteJunction');
  });

  it('should use number type for ID if the id field type is number', async () => {
    const model: ParsedModel = {
      name: 'Product',
      dbName: null,
      fields: [
        {
          name: 'id',
          type: 'number',
          kind: 'scalar' as DMMF.FieldKind,
          isRequired: true,
          isId: true,
          isUnique: true,
          isUpdatedAt: false,
          isList: false,
          hasDefaultValue: true,
        },
      ],
    };

    const result = await generateServiceFileContent(model);
    expect(result).toMatchSnapshot();
    // ID parameter should be typed as number
    expect(result).toContain('id: number');
  });

  it('should generate service file content with a custom prisma client path', async () => {
    const model: ParsedModel = {
      name: 'User',
      dbName: null,
      fields: [
        {
          name: 'id',
          type: 'string',
          kind: 'scalar' as DMMF.FieldKind,
          isRequired: true,
          isId: true,
          isUnique: true,
          isUpdatedAt: false,
          isList: false,
          hasDefaultValue: true,
        },
      ],
    };

    const customPath = '@prisma/custom-client';
    const result = await generateServiceFileContent(model, customPath);
    expect(result).toMatchSnapshot();
    expect(result).toContain(
      `import type { PrismaClient } from '${customPath}'`
    );
  });

  it('should preserve underscores in model names when generating prisma client access', async () => {
    const model: ParsedModel = {
      name: 'Hist_AuditoriaDesarrollador',
      dbName: null,
      fields: [
        {
          name: 'id',
          type: 'string',
          kind: 'scalar' as DMMF.FieldKind,
          isRequired: true,
          isId: true,
          isUnique: true,
          isUpdatedAt: false,
          isList: false,
          hasDefaultValue: true,
        },
      ],
    };

    const result = await generateServiceFileContent(model);
    expect(result).toMatchSnapshot();

    // Function names should use PascalCase without underscores
    expect(result).toContain('export const findManyHistAuditoriaDesarrollador');
    expect(result).toContain('export const createHistAuditoriaDesarrollador');

    // Prisma client access should preserve the underscore
    expect(result).toContain('prisma.hist_AuditoriaDesarrollador.findMany');
    expect(result).toContain('prisma.hist_AuditoriaDesarrollador.create');
    expect(result).toContain('prisma.hist_AuditoriaDesarrollador.findUnique');
    expect(result).toContain('prisma.hist_AuditoriaDesarrollador.update');
    expect(result).toContain('prisma.hist_AuditoriaDesarrollador.delete');
  });
});
