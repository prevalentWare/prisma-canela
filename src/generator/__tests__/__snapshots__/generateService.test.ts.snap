// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateServiceFileContent > should generate correct service file content for a User model 1`] = `
"
import type { PrismaClient } from 'prisma/schema/generated/client';
// We need to import the actual model type separately when verbatimModuleSyntax is true
import type { User as UserType } from 'prisma/schema/generated/client';
// Import input types from Zod schemas (adjust path if necessary)
import { createUserSchema, updateUserSchema } from './schema';
import { z } from 'zod';
import type { Context } from 'hono';

// Define types inferred from Zod schemas for input validation
type CreateUserInput = z.infer<typeof createUserSchema>;
type UpdateUserInput = z.infer<typeof updateUserSchema>;

/**
 * Helper function to extract Prisma client from context with error handling
 */
const getPrismaClient = (c: Context): PrismaClient => {
  const prisma = c.get('prisma');
  if (!prisma) {
    throw new Error('Prisma client not found in context. Make sure to use the prismaMiddleware.');
  }
  return prisma;
};

/**
 * Finds multiple User records.
 * @param c The Hono context containing the Prisma client.
 * @returns A promise resolving to an array of User records.
 */
export const findManyUser = async (c: Context): Promise<UserType[]> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.findMany();
  } catch (error) {
    console.error('Error fetching Users:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch Users');
  }
};

/**
 * Creates a new User record.
 * @param c The Hono context containing the Prisma client.
 * @param data The data for the new User.
 * @returns A promise resolving to the created User record.
 */
export const createUser = async (
  c: Context, 
  data: CreateUserInput
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.user.create({ data });
  } catch (error) {
    console.error('Error creating User:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not create User');
  }
};

/**
 * Finds a single User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to find.
 * @returns A promise resolving to the User record or null if not found.
 */
export const findUserById = async (
  c: Context, 
  id: string
): Promise<UserType | null> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.findUnique({
      where: { id: id },
    });
  } catch (error) {
    console.error(\`Error fetching User by ID \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch User by ID');
  }
};

/**
 * Updates a User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to update.
 * @param data The data to update the User with.
 * @returns A promise resolving to the updated User record.
 */
export const updateUser = async (
  c: Context, 
  id: string, 
  data: UpdateUserInput
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.user.update({
      where: { id: id },
      data,
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error updating User \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not update User');
  }
};

/**
 * Deletes a User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to delete.
 * @returns A promise resolving to the deleted User record.
 */
export const deleteUser = async (
  c: Context, 
  id: string
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.delete({
      where: { id: id },
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error deleting User \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not delete User');
  }
};
"
`;

exports[`generateServiceFileContent > should generate service file content for a model without an ID 1`] = `
"
import type { PrismaClient } from 'prisma/schema/generated/client';
// We need to import the actual model type separately when verbatimModuleSyntax is true
import type { Junction as JunctionType } from 'prisma/schema/generated/client';
// Import input types from Zod schemas (adjust path if necessary)
import { createJunctionSchema, updateJunctionSchema } from './schema';
import { z } from 'zod';
import type { Context } from 'hono';

// Define types inferred from Zod schemas for input validation
type CreateJunctionInput = z.infer<typeof createJunctionSchema>;
type UpdateJunctionInput = z.infer<typeof updateJunctionSchema>;

/**
 * Helper function to extract Prisma client from context with error handling
 */
const getPrismaClient = (c: Context): PrismaClient => {
  const prisma = c.get('prisma');
  if (!prisma) {
    throw new Error('Prisma client not found in context. Make sure to use the prismaMiddleware.');
  }
  return prisma;
};

/**
 * Finds multiple Junction records.
 * @param c The Hono context containing the Prisma client.
 * @returns A promise resolving to an array of Junction records.
 */
export const findManyJunction = async (c: Context): Promise<JunctionType[]> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.junction.findMany();
  } catch (error) {
    console.error('Error fetching Junctions:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch Junctions');
  }
};

/**
 * Creates a new Junction record.
 * @param c The Hono context containing the Prisma client.
 * @param data The data for the new Junction.
 * @returns A promise resolving to the created Junction record.
 */
export const createJunction = async (
  c: Context, 
  data: CreateJunctionInput
): Promise<JunctionType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.junction.create({ data });
  } catch (error) {
    console.error('Error creating Junction:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not create Junction');
  }
};
"
`;

exports[`generateServiceFileContent > should generate service file content with a custom prisma client path 1`] = `
"
import type { PrismaClient } from '../../libs/prisma';
// We need to import the actual model type separately when verbatimModuleSyntax is true
import type { User as UserType } from '../../libs/prisma';
// Import input types from Zod schemas (adjust path if necessary)
import { createUserSchema, updateUserSchema } from './schema';
import { z } from 'zod';
import type { Context } from 'hono';

// Define types inferred from Zod schemas for input validation
type CreateUserInput = z.infer<typeof createUserSchema>;
type UpdateUserInput = z.infer<typeof updateUserSchema>;

/**
 * Helper function to extract Prisma client from context with error handling
 */
const getPrismaClient = (c: Context): PrismaClient => {
  const prisma = c.get('prisma');
  if (!prisma) {
    throw new Error('Prisma client not found in context. Make sure to use the prismaMiddleware.');
  }
  return prisma;
};

/**
 * Finds multiple User records.
 * @param c The Hono context containing the Prisma client.
 * @returns A promise resolving to an array of User records.
 */
export const findManyUser = async (c: Context): Promise<UserType[]> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.findMany();
  } catch (error) {
    console.error('Error fetching Users:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch Users');
  }
};

/**
 * Creates a new User record.
 * @param c The Hono context containing the Prisma client.
 * @param data The data for the new User.
 * @returns A promise resolving to the created User record.
 */
export const createUser = async (
  c: Context, 
  data: CreateUserInput
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.user.create({ data });
  } catch (error) {
    console.error('Error creating User:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not create User');
  }
};

/**
 * Finds a single User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to find.
 * @returns A promise resolving to the User record or null if not found.
 */
export const findUserById = async (
  c: Context, 
  id: string
): Promise<UserType | null> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.findUnique({
      where: { id: id },
    });
  } catch (error) {
    console.error(\`Error fetching User by ID \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch User by ID');
  }
};

/**
 * Updates a User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to update.
 * @param data The data to update the User with.
 * @returns A promise resolving to the updated User record.
 */
export const updateUser = async (
  c: Context, 
  id: string, 
  data: UpdateUserInput
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.user.update({
      where: { id: id },
      data,
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error updating User \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not update User');
  }
};

/**
 * Deletes a User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to delete.
 * @returns A promise resolving to the deleted User record.
 */
export const deleteUser = async (
  c: Context, 
  id: string
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.delete({
      where: { id: id },
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error deleting User \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not delete User');
  }
};
"
`;

exports[`generateServiceFileContent > should generate service file content with a custom prisma client path 2`] = `
"
import type { PrismaClient } from '@prisma/custom-client';
// We need to import the actual model type separately when verbatimModuleSyntax is true
import type { User as UserType } from '@prisma/custom-client';
// Import input types from Zod schemas (adjust path if necessary)
import { createUserSchema, updateUserSchema } from './schema';
import { z } from 'zod';
import type { Context } from 'hono';

// Define types inferred from Zod schemas for input validation
type CreateUserInput = z.infer<typeof createUserSchema>;
type UpdateUserInput = z.infer<typeof updateUserSchema>;

/**
 * Helper function to extract Prisma client from context with error handling
 */
const getPrismaClient = (c: Context): PrismaClient => {
  const prisma = c.get('prisma');
  if (!prisma) {
    throw new Error('Prisma client not found in context. Make sure to use the prismaMiddleware.');
  }
  return prisma;
};

/**
 * Finds multiple User records.
 * @param c The Hono context containing the Prisma client.
 * @returns A promise resolving to an array of User records.
 */
export const findManyUser = async (c: Context): Promise<UserType[]> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.findMany();
  } catch (error) {
    console.error('Error fetching Users:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch Users');
  }
};

/**
 * Creates a new User record.
 * @param c The Hono context containing the Prisma client.
 * @param data The data for the new User.
 * @returns A promise resolving to the created User record.
 */
export const createUser = async (
  c: Context, 
  data: CreateUserInput
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.user.create({ data });
  } catch (error) {
    console.error('Error creating User:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not create User');
  }
};

/**
 * Finds a single User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to find.
 * @returns A promise resolving to the User record or null if not found.
 */
export const findUserById = async (
  c: Context, 
  id: string
): Promise<UserType | null> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.findUnique({
      where: { id: id },
    });
  } catch (error) {
    console.error(\`Error fetching User by ID \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch User by ID');
  }
};

/**
 * Updates a User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to update.
 * @param data The data to update the User with.
 * @returns A promise resolving to the updated User record.
 */
export const updateUser = async (
  c: Context, 
  id: string, 
  data: UpdateUserInput
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.user.update({
      where: { id: id },
      data,
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error updating User \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not update User');
  }
};

/**
 * Deletes a User record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the User to delete.
 * @returns A promise resolving to the deleted User record.
 */
export const deleteUser = async (
  c: Context, 
  id: string
): Promise<UserType> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.user.delete({
      where: { id: id },
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error deleting User \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not delete User');
  }
};
"
`;

exports[`generateServiceFileContent > should preserve underscores in model names when generating prisma client access 1`] = `
"
import type { PrismaClient } from 'prisma/schema/generated/client';
// We need to import the actual model type separately when verbatimModuleSyntax is true
import type { Hist_AuditoriaDesarrollador as HistAuditoriaDesarrolladorType } from 'prisma/schema/generated/client';
// Import input types from Zod schemas (adjust path if necessary)
import { createHistAuditoriaDesarrolladorSchema, updateHistAuditoriaDesarrolladorSchema } from './schema';
import { z } from 'zod';
import type { Context } from 'hono';

// Define types inferred from Zod schemas for input validation
type CreateHistAuditoriaDesarrolladorInput = z.infer<typeof createHistAuditoriaDesarrolladorSchema>;
type UpdateHistAuditoriaDesarrolladorInput = z.infer<typeof updateHistAuditoriaDesarrolladorSchema>;

/**
 * Helper function to extract Prisma client from context with error handling
 */
const getPrismaClient = (c: Context): PrismaClient => {
  const prisma = c.get('prisma');
  if (!prisma) {
    throw new Error('Prisma client not found in context. Make sure to use the prismaMiddleware.');
  }
  return prisma;
};

/**
 * Finds multiple HistAuditoriaDesarrollador records.
 * @param c The Hono context containing the Prisma client.
 * @returns A promise resolving to an array of HistAuditoriaDesarrollador records.
 */
export const findManyHistAuditoriaDesarrollador = async (c: Context): Promise<HistAuditoriaDesarrolladorType[]> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.hist_AuditoriaDesarrollador.findMany();
  } catch (error) {
    console.error('Error fetching HistAuditoriaDesarrolladors:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch HistAuditoriaDesarrolladors');
  }
};

/**
 * Creates a new HistAuditoriaDesarrollador record.
 * @param c The Hono context containing the Prisma client.
 * @param data The data for the new HistAuditoriaDesarrollador.
 * @returns A promise resolving to the created HistAuditoriaDesarrollador record.
 */
export const createHistAuditoriaDesarrollador = async (
  c: Context, 
  data: CreateHistAuditoriaDesarrolladorInput
): Promise<HistAuditoriaDesarrolladorType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.hist_AuditoriaDesarrollador.create({ data });
  } catch (error) {
    console.error('Error creating HistAuditoriaDesarrollador:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not create HistAuditoriaDesarrollador');
  }
};

/**
 * Finds a single HistAuditoriaDesarrollador record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the HistAuditoriaDesarrollador to find.
 * @returns A promise resolving to the HistAuditoriaDesarrollador record or null if not found.
 */
export const findHistAuditoriaDesarrolladorById = async (
  c: Context, 
  id: string
): Promise<HistAuditoriaDesarrolladorType | null> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.hist_AuditoriaDesarrollador.findUnique({
      where: { id: id },
    });
  } catch (error) {
    console.error(\`Error fetching HistAuditoriaDesarrollador by ID \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch HistAuditoriaDesarrollador by ID');
  }
};

/**
 * Updates a HistAuditoriaDesarrollador record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the HistAuditoriaDesarrollador to update.
 * @param data The data to update the HistAuditoriaDesarrollador with.
 * @returns A promise resolving to the updated HistAuditoriaDesarrollador record.
 */
export const updateHistAuditoriaDesarrollador = async (
  c: Context, 
  id: string, 
  data: UpdateHistAuditoriaDesarrolladorInput
): Promise<HistAuditoriaDesarrolladorType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.hist_AuditoriaDesarrollador.update({
      where: { id: id },
      data,
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error updating HistAuditoriaDesarrollador \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not update HistAuditoriaDesarrollador');
  }
};

/**
 * Deletes a HistAuditoriaDesarrollador record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the HistAuditoriaDesarrollador to delete.
 * @returns A promise resolving to the deleted HistAuditoriaDesarrollador record.
 */
export const deleteHistAuditoriaDesarrollador = async (
  c: Context, 
  id: string
): Promise<HistAuditoriaDesarrolladorType> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.hist_AuditoriaDesarrollador.delete({
      where: { id: id },
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error deleting HistAuditoriaDesarrollador \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not delete HistAuditoriaDesarrollador');
  }
};
"
`;

exports[`generateServiceFileContent > should use number type for ID if the id field type is number 1`] = `
"
import type { PrismaClient } from 'prisma/schema/generated/client';
// We need to import the actual model type separately when verbatimModuleSyntax is true
import type { Product as ProductType } from 'prisma/schema/generated/client';
// Import input types from Zod schemas (adjust path if necessary)
import { createProductSchema, updateProductSchema } from './schema';
import { z } from 'zod';
import type { Context } from 'hono';

// Define types inferred from Zod schemas for input validation
type CreateProductInput = z.infer<typeof createProductSchema>;
type UpdateProductInput = z.infer<typeof updateProductSchema>;

/**
 * Helper function to extract Prisma client from context with error handling
 */
const getPrismaClient = (c: Context): PrismaClient => {
  const prisma = c.get('prisma');
  if (!prisma) {
    throw new Error('Prisma client not found in context. Make sure to use the prismaMiddleware.');
  }
  return prisma;
};

/**
 * Finds multiple Product records.
 * @param c The Hono context containing the Prisma client.
 * @returns A promise resolving to an array of Product records.
 */
export const findManyProduct = async (c: Context): Promise<ProductType[]> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.product.findMany();
  } catch (error) {
    console.error('Error fetching Products:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch Products');
  }
};

/**
 * Creates a new Product record.
 * @param c The Hono context containing the Prisma client.
 * @param data The data for the new Product.
 * @returns A promise resolving to the created Product record.
 */
export const createProduct = async (
  c: Context, 
  data: CreateProductInput
): Promise<ProductType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.product.create({ data });
  } catch (error) {
    console.error('Error creating Product:', error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not create Product');
  }
};

/**
 * Finds a single Product record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the Product to find.
 * @returns A promise resolving to the Product record or null if not found.
 */
export const findProductById = async (
  c: Context, 
  id: number
): Promise<ProductType | null> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.product.findUnique({
      where: { id: id },
    });
  } catch (error) {
    console.error(\`Error fetching Product by ID \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not fetch Product by ID');
  }
};

/**
 * Updates a Product record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the Product to update.
 * @param data The data to update the Product with.
 * @returns A promise resolving to the updated Product record.
 */
export const updateProduct = async (
  c: Context, 
  id: number, 
  data: UpdateProductInput
): Promise<ProductType> => {
  try {
    const prisma = getPrismaClient(c);
    // Input data is assumed to be validated by Zod schema in the route handler
    return await prisma.product.update({
      where: { id: id },
      data,
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error updating Product \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not update Product');
  }
};

/**
 * Deletes a Product record by its ID.
 * @param c The Hono context containing the Prisma client.
 * @param id The ID of the Product to delete.
 * @returns A promise resolving to the deleted Product record.
 */
export const deleteProduct = async (
  c: Context, 
  id: number
): Promise<ProductType> => {
  try {
    const prisma = getPrismaClient(c);
    return await prisma.product.delete({
      where: { id: id },
    });
  } catch (error) {
    // TODO: Differentiate between Prisma's P2025 error (Record not found) and other errors
    console.error(\`Error deleting Product \${id}:\`, error);
    // TODO: Implement more specific error handling and logging
    throw new Error('Could not delete Product');
  }
};
"
`;
